File: victor-clad/tools/message-buffers/emitters/tests/javascript/clad_test.js
const { TestClad } = require('./build/simple/Javascript.js');
const assert = require('assert');
const fs = require('fs');

// Tests for Javascript CLAD emitter 

let testSubEnum = TestClad.EnumT.One;
let testSubBool = false;

let testSubMessage = new TestClad.SubMessage(testSubBool, testSubEnum);

// cladMessage fields
let testBool = true;
let testUint8 = 147;
let testUint16 = 35149;
let testUint32 = 1290345;
let testUint64 = 5239140128n;
let testFloat32 = 343.2;
let testFloat64 = 12999.1234;
let testString = "Hello, world!";
let testString16 = "Hi, a string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. The End.";
let testFArray8 = new Uint8Array([ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27 ]);
let testVArray8 = new Uint8Array([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ]);
let testVArrayFloat32 = new Float32Array([1.23, 2.34, 3.45, 4.56, 5.67, 6.78, 7.89]);
let testStringArray = [ "First", "Second", "Third", "Fourth", "Fifth", "Sixth" ];

let cladMessage = new TestClad.Message(
  testBool, 
  testUint8,
  testUint16,
  testUint32, 
  testUint64,
  testFloat32, 
  testFloat64, 
  testSubMessage, 
  testString, 
  testString16, 
  testFArray8, 
  testVArray8, 
  testVArrayFloat32,
  testStringArray);

let buffer = cladMessage.pack();

// write buffer to file for interop testing
fs.writeFileSync("./buffer.tmp", Buffer.from(buffer), function(err) {
  //
  console.log("write file");
});

// scramble data
cladMessage.testBool = null;
cladMessage.testUint8 = null;
cladMessage.testUint16 = null;
cladMessage.testUint32 = null;
cladMessage.testUint64 = null;
cladMessage.testFloat32 = null;
cladMessage.testFloat64 = null;
cladMessage.testSubMessage = null;
cladMessage.testString = null;
cladMessage.testString16 = null;
cladMessage.testFArray8 = null;
cladMessage.testVArray8 = null;
cladMessage.testVArrayFloat32 = null;
cladMessage.testStringArray = null;

cladMessage.unpack(buffer);

assert(testBool == cladMessage.testBool, "bool test failed.");
assert(testUint8 == cladMessage.testUint8, "uint_8 test failed.");
assert(testUint16 == cladMessage.testUint16, "uint_16 test failed.");
assert(testUint32 == cladMessage.testUint32, "uint_32 test failed.");
assert(testUint64 == cladMessage.testUint64, "uint_64 test failed.");
assert(Math.abs(testFloat32 - cladMessage.testFloat32) < 0.001, "float_32 test failed.");
assert(Math.abs(testFloat64 - cladMessage.testFloat64) < 0.001, "float_64 test failed.");
assert(testSubMessage.testSubBool == cladMessage.testSubMessage.testSubBool, "Bool field of sub-message test failed.");
assert(testSubMessage.testSubEnum == cladMessage.testSubMessage.testSubEnum, "Enum field of sub-message test failed.");
assert(testString == cladMessage.testString, "String with max length MAX_UINT8 test failed.");
assert(testString16 == cladMessage.testString16, "String with max length MAX_UINT16 test failed.");

function checkArrayEquality(a, b) {
  if(a === b) return true;
  if(a == null || b == null) return false;
  if(a.length != b.length) return false;

  for(let i = 0; i < a.length; i++) {
    if(a[i] != b[i]) return false;
  }

  return true;
}

function checkArrayFloatProximity(a, b) {
  if(a === b) return true;
  if(a == null || b == null) return false;
  if(a.length != b.length) return false;

  for(let i = 0; i < a.length; i++) {
    if(Math.abs(a[i] - b[i]) >= 0.001) return false;
  }

  return true;
}

assert(checkArrayEquality(testFArray8, cladMessage.testFArray8), "Fixed array of uint8s test failed.");
assert(checkArrayEquality(testVArray8, cladMessage.testVArray8), "Variable array of uint8s test failed.");
assert(checkArrayFloatProximity(testVArrayFloat32, cladMessage.testVArrayFloat32), "Array of floats test failed.");
assert(checkArrayEquality(testStringArray, cladMessage.testStringArray), "Array of strings test failed.");

console.log("All tests passed!");

File: victor-clad/tools/message-buffers/emitters/tests/javascript/clad_test.cpp
#include "build/simple/JavascriptTag.h"
#include "build/simple/Javascript.h"
#include <stdio.h>
#include <fstream>
#include <vector>
#include <cmath>

const static std::string sTestString16 = "Hi, a string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. A string of length longer than 256 characters. The End.";
const static uint8_t sTestFArray8[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27 };
const static uint8_t sTestVArray8[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
const static float sTestVArrayFloat[] = { 1.23f, 2.34f, 3.45f, 4.56f, 5.67f, 6.78f, 7.89f };
const static std::string sTestStringArray[] = { "First", "Second", "Third", "Fourth", "Fifth", "Sixth" };

template<class T>
bool ArrayMatches(std::vector<T> a, std::vector<T> b) {
  if(a.size() != b.size()) {
    return false;
  }

  for(int i = 0; i < a.size(); i++) {
    if(a[i] != b[i]) {
      return false;
    }
  }

  return true;
}

template<>
bool ArrayMatches<float>(std::vector<float> a, std::vector<float> b) {
  if(a.size() != b.size()) {
    return false;
  }

  for(int i = 0; i < a.size(); i++) {
    if(std::abs(float(a[i] - b[i])) >= 0.001f) {
      return false;
    }
  }

  return true;
}

template<class T>
bool ArrayMatches(std::array<T, 27> a, std::vector<T> b) {
  if(a.size() != b.size()) {
    return false;
  }

  for(int i = 0; i < a.size(); i++) {
    if(a[i] != b[i]) {
      return false;
    }
  }

  return true;
}

int main() {
  std::ifstream f("buffer.tmp");

  f.seekg(0, f.end);
  size_t size = f.tellg();
  f.seekg(0, f.beg);

  char buffer[2048];

  if(size > sizeof(buffer)) {
    size = sizeof(buffer);
  }

  f.read(buffer, size);

  TestClad::Message msg;

  msg.Unpack((uint8_t*)buffer, size);

  std::vector<std::string> testStringVector;
  for(int i = 0; i < sizeof(sTestStringArray) / sizeof(std::string); i++) {
    testStringVector.push_back(sTestStringArray[i]);
  }

  assert(msg.testBool == true);
  assert(msg.testUint8 == 147);
  assert(msg.testUint16 == 35149);
  assert(msg.testUint32 == 1290345);
  assert(msg.testUint64 == 5239140128);
  assert(std::abs(float(msg.testFloat32 - 343.2f)) < 0.001f);
  assert(std::abs(float(msg.testFloat64 - 12999.1234f)) < 0.001f);
  assert(msg.testString == "Hello, world!");
  assert(msg.testString16 == sTestString16);
  assert(ArrayMatches(msg.testFArray8, std::vector<uint8_t>(sTestFArray8, sTestFArray8 + sizeof(sTestFArray8))));
  assert(ArrayMatches(msg.testVArray8, std::vector<uint8_t>(sTestVArray8, sTestVArray8 + sizeof(sTestVArray8))));
  assert(ArrayMatches(msg.testStringArray, testStringVector));
  assert(ArrayMatches(msg.testVArrayFloat32, std::vector<float>(sTestVArrayFloat, sTestVArrayFloat + sizeof(sTestVArrayFloat) / sizeof(float))));
  printf("Test passed: JavaScript emitter interop with CPP emitter.\n");

  return 0;
}